service-base/src/main/java/org/eclipse/hono/service/resourcelimits/ResourceLimitChecks.java

```diff
@@ -56,6 +57,22 @@
      */
     Future<Boolean> isMessageLimitReached(TenantObject tenantObject, long payloadSize, SpanContext spanContext);
 
+    /**
+     * Adds the given amount to the cached value of consumed bytes for the given tenant.
+     * <br>
+     * By using this method the accuracy of the message limit can be improved because the cached value for consumed bytes
+     * is almost instantly stale after it was queried by Prometheus as more traffic flows in.
+     *
+     * @param tenant The tenant for which the given payload should be added to the cache.
+     * @param payloadSize The amount of bytes to be added to the cache.
+     * @param outcome The outcome of the operation used for uploading the given payload. The cache will not be updated
+     *                for {@link org.eclipse.hono.service.metric.MetricsTags.ProcessingOutcome#UNDELIVERABLE} outcomes
+     *                as this is not the client's fault and shouldn't add to the number of consumed bytes.
+     *
+     * @return A future indicating when the update was done.
+     */
+    Future<Void> addToMessageBytesCache(String tenant, long payloadSize, MetricsTags.ProcessingOutcome outcome);
```

Comment:
how about putting this method into its own interface and then inject an implementation into `MicrometerBasedMetrics` so that you can update the local cache as part of the telemetry reporting. We would then also not need to change every protocol adapter and make sure to not only report metrics via Micrometer but also invoke the ResourceLimitChecks ...

Reply:
Sound reasonable. But I think we also need to include it as part of `reportCommand`, since the data volume resource limits also apply for commands, right?

Reply:
yes

Reply:
Done. Let me know what you think.